experimental/const_provider/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+                            key: key.clone(),
+                            options: options,
+                        },
+                    };
+                    let payload: DataPayload<$marker> =
+                        $provider.load_payload(&req).expect("Payload").take_payload().expect("Payload");
+                    let payload_bake = payload.get().bake();
+                    quote! { (#options_bake, #payload_bake) }
+                });
+
+                let field_name = format_ident!("{}_{}_{}", key.category.to_string(), &*key.sub_category, key.version);
+
+                imports.push(quote! {
+                    use $data_struct;
+                });
+
+                field_declarations.push(quote! {
+                    #field_name: &'static [(icu_provider::prelude::ResourceOptions, $data_struct<'static>)],
+                });
+
+                field_values.push(quote! {
+                    #field_name: &[#(#data),*],
+                });
+
+                let category = key.category.bake();
+                let sub_category = key.sub_category.as_unsigned();
+                let version = key.version;
+                keys_to_fields_match_arms.push(quote! {
+                    (#category, #sub_category, #version) => self.#field_name,
+                });
+            }
+
+            data_provider_impls.push(quote! {
+                impl icu_provider::prelude::DataProvider<$marker> for ConstProvider {
+                    fn load_payload(&self, req: &icu_provider::prelude::DataRequest) ->
+                            Result<icu_provider::prelude::DataResponse<$marker>, icu_provider::prelude::DataError> {
+                        use icu_provider::prelude::*;
+                        use icu_provider::yoke::ZeroCopyFrom;
+
+                        let data: &[(ResourceOptions, $data_struct)] =
+                            match (
+                                req.resource_path.key.category,
+                                req.resource_path.key.sub_category.as_unsigned(),
+                                req.resource_path.key.version) {
+                                #(#keys_to_fields_match_arms)*
+                                _ => return Err(DataError::MissingResourceKey(req.resource_path.key.clone())),
+                            };
+                        #[allow(unreachable_code)] // If no key is included
+                        match data.binary_search_by_key(&&req.resource_path.options, |(k,_)| k) {
+                            Err(_) => return Err(DataError::MissingResourceOptions(req.clone())),
+                            Ok(index) => {
+                                // TODO(#1109): Set metadata.data_langid correctly.
+                                Ok(DataResponse {
+                                    metadata: DataResponseMetadata::default(),
+                                    payload: Some(DataPayload::from_owned(<$data_struct>::zero_copy_from(&data[index].1))),
```

Comment:
issue: this can just be `.clone()`, it's a shallow pointer copy for zero copy stuff